### PR TITLE
MLX backend: System One reads on Apple silicon

### PR DESCRIPTION
**Pull request by gatewaynode**
OPENJEV_BACKEND=mlx runs DiffusionGemma in-process through MLX and mlx-vlm: no vLLM, no Docker. The engine's schema, template, canvas and seeds are unchanged; MlxEngine overrides only the read. Text reads, samples, sequential and automatic re-reads work. images, think and steps above 1 get a 422, and /v1/chat/completions a 501, until they are ported. The default vLLM path is unchanged.

### CODE REVIEW

**Comment by razorback16** [comment]
Merged — thanks for this. Some notes from reviewing and running it, since it landed without a round-trip.

**It works.** I ran the backend on an M-series Mac against `mlx-community/diffusiongemma-26B-A4B-it-4bit`: the model loads in 2.8 s and a three-question read returns in 1.3 s with the right answers (urgent 0.997, team `outage` at p=0.99999, tone 2.0). All six tests in `tests/test_mlx_model.py` pass against real weights.

The parts I expected to be shaky hold up:

- The three `diffusion_*` methods exist in `mlx-vlm==0.6.15` with the signatures you call.
- The prefill reuse is safe. In `decoder=True` mode the attention reads `_cache_state(cache)`, sets `attn_cache = None` and never calls `update_and_fetch`, so a decoder pass leaves the cache as it found it. Your `test_a_cached_prefill_reads_the_same` passes bitwise against real weights — good call covering that with a real-weights test rather than an argument.
- The default repo ships a `chat_template.jinja` byte-identical to the nvidia NVFP4 one (same SHA-256, 17336 bytes), so prompts and the single-token label search come out the same on either tokenizer.

**One thing you could not have seen from your branch.** It was based on `91d5005` and predated `c47a1de`, which replaced `validation_error` with `semantic_error`: a `SchemaError` is now **400 with a plain-string `detail`**, not 422 with a list. GitHub reported MERGEABLE because there was no textual conflict, but after merging `main` three tests failed — `test_unsupported_options_are_refused`, `test_long_prompts_are_refused` and `test_mlx_model.py::test_refusals` — all asserting 422 and indexing `detail[0]["loc"]`, which is a `TypeError` on a string.

The behaviour was already right; `MlxEngine.decide` raises `SchemaError` with the field name leading the message, which is exactly what main's own test expects. I fixed the assertions and docs in `e4098ff`. Worth knowing for next time: rebase on `main` before the tests tell you something surprising.

Two things I did not change, because they are design calls rather than defects:

**1. Label ids are selected per slot, not across the request.** `Engine.one_read` sends `sorted({i for s in slots for i in s["label_ids"]})[:MAX_LABEL_IDS]` — the union across *all* slots, requested at every position (`openjev/engine.py:250`) — while `MlxRuntime.read` keeps only the current slot's labels (`openjev/mlx_backend.py:68`). `slot_distribution` then takes `floor = min(top.values()) - 5.0` and computes `entropy` over the whole `top` dict, so for identical logits both the floor and the entropy driving `auto_threshold` re-reads differ between the two backends. The effect is small — the extra ids are low-probability — but it is a divergence in a change whose premise is that only the read differs. Hoisting the union once in `MlxEngine.one_read` and passing it into `read` is a one-liner.

**2. `MlxRuntime` has no coverage off Apple silicon.** `monkeypatch.setattr(mlx_backend, "MlxRuntime", StubRuntime)` swaps out the whole class, so the prefill cache and its eviction, the top-k ∪ labels selection and the log-softmax never execute in CI; `tests/test_mlx_model.py` covers them but skips without `OPENJEV_MLX_TEST_MODEL`. I drove `MlxRuntime` against a fake `model` exposing the three `diffusion_*` methods and it is all correct — cache reuse, 23 kept ids for TOPK=20 with 3 labels, normalised log-softmax, eviction bounding at 16200 ≤ 16384 — so this is about pinning the behaviour, not fixing it. The eviction loop especially:

```python
while len(self.prefills) > 1 and sum(map(len, self.prefills)) > PROMPT_CACHE_TOKENS:
```

`map(len, self.prefills)` sums over the *keys* — the prompt-token tuples — which is the point, and exactly the kind of line someone later "fixes" to `.values()`. A test pins the intent. (I added a comment saying so, but a test would be better.)

Happy to take either as a follow-up PR if you want them.
